### DISCUSSION TITLE
Any support for Wayland?

### DISCUSSION THREAD

**Discussion by spit4520** [FreeSpacenav/spacenavd#37] [Questions and support]
I am on Ubuntu 20.04 and having issues when not running in X11 comp mode. 

Any thoughts?

Thanks

S

**Comment by jtsiomb** [comment]
What does wayland have to do with spacenavd? What kind of support would you like?

**Comment by Unisa-Merni-392** [comment]
well, it complains that it can't connect to display ":0", for starters.
I don't know what's the deal here, since I run sway with xwayland, so theoretically it should work, but it doesn't

**Reply by jtsiomb** [reply]
If you're running an X server and spacenavd fails to connect to it, the most likely culprit is that you don't have the correct authentication cookie where Xlib expects it to be. See FAQ#10: https://spacenav.sourceforge.net/faq.html#faq10

Another possibility is that your X server is not running as display ":0" and you don't have the DISPLAY environment variable set up to whatever display it's running as.

**Reply by Unisa-Merni-392** [reply]
As I stated, I'm not running X11/Xorg directly. 
I'm running sway on Wayland and it has xwayland support for embedded Xorg server that should serve apps that want to talk to Xorg.

I don't have .Xauthority file in home dir as it looks like spacenavd can't reach or talk to my Xorg server that runs within wayland.

Given that majority of distros are migrating to Wayland these days (with MANY good reasons!), this seems as something that should be looked into...

**Reply by jtsiomb** [reply]
Regardless of how the X server is running, be it standalone or rootless on some other window system, applications still need to authenticate to connect to it, and certainly it will be running as a certain display number, and that number will have to be specified in the DISPLAY environment variable. There might be some other cause to your problem, but these two are the only ones I can think of right now. Here's a list of questions, and things to try:

  - Are other X apps running on your setup?
  - Is `DISPLAY` set, and to what value?
  - How are you running spacenavd, from init as a daemon or interactively in an X terminal? Try running it manually and see if something changes: `spacenavd -v -d` as root.
  - If authentication is the issue, running ~~xauth +~~ `xhost +` should disable X authentication and allow all clients to connect, does that change anything?

For the sake of expediency, I'll try to take a blind stab at a probable cause before waiting for your answers:

I assume you're running spacenavd from init (as you should, it is a daemon after all). I also assume (because I remember reading something to that effect by another wayland user) that wayland is not running the X server as display 0 which is the usual case, but as some larger number instead. During init the DISPLAY environment variable is probably not set, since the X server is not running yet, which causes spacenavd to fallback to assuming display 0. When the X server starts as a different display number spacenavd fails to connect to it.

If the above assumption turns out to be correct, and further assuming the display number X starts as is predictable and doesn't change from one session to the next, you can edit your init script and make sure to manually define `DISPLAY` before starting spacenavd. Also if the above turns out to be the issue, I encourage you to add a feature request in the issue tracker, to improve X server display number detection. Otherwise it's likely to be forgotten in a few weeks.

**Reply by Unisa-Merni-392** [reply]
OK, here we go:

INFO:I'm running Gentoo, with wayland+sway with enabled xwayland support.
INFO2: I've read GEntoo's page on Space Navigator: https://wiki.gentoo.org/wiki/SpaceNavigator

AD 1: Yes. I've just installed sci-electronics/pcb as test example. Runs fine with ("xwayland on") in sway config file, it doesn't start with ("xwayland off").
AD 2: DISPLAY=":0"
AD3: Spacenavd is being stared from systemd through sxpacenavd.service. It executes it as /usr/bin/spacenavd -d 
I've tried running it directly as root and reulsts were the same. Tired running it also as user, owning the sesssion, but even after relaxing the permissions for /var/run, it didn't work
AD 4: "xauth +" doesn't work for me. I get "unknown command +" response.

Notice:
I never have ~/.Xauthority file. Not even when running pcb and it connects to xwayland.

**Reply by jtsiomb** [reply]
I apologize, my mistake, the command is `xhost +`, not xauth. Try running that and then run spacenavd manually.

**Reply by Unisa-Merni-392** [reply]
I tried that and now spacenavd stops complaining about authorization.
But still neither test_cube example nor freecad react to the device.

So I've ran it manually and it seems that the daemon can't find the device it would recognise:
"/usr/bin/spacenavd -d -vvvv" gives me:

***********************************************************
Spacenav daemon 1.2
reading config file: /etc/spnavrc
Device detection, parsing /proc/bus/input/devices
trying alternative detection, querying /dev/input/ devices...
  trying "/dev/input/." ... 
  trying "/dev/input/.." ... 
  trying "/dev/input/event14" ... 
  trying "/dev/input/event13" ... 
  trying "/dev/input/event12" ... 
  trying "/dev/input/by-id" ... 
  trying "/dev/input/by-path" ... 
  trying "/dev/input/event10" ... 
  trying "/dev/input/mouse0" ... 
  trying "/dev/input/event9" ... 
  trying "/dev/input/event8" ... 
  trying "/dev/input/event7" ... 
  trying "/dev/input/event6" ... 
  trying "/dev/input/event5" ... 
  trying "/dev/input/event4" ... 
  trying "/dev/input/event3" ... 
  trying "/dev/input/event2" ... 
  trying "/dev/input/event1" ... 
  trying "/dev/input/event0" ... 
  trying "/dev/input/mice" ... 
failed to find any supported USB devices
trying to open X11 display ":0.0"
   XAUTHORITY=/root/.Xauthority
Using XTEST to send key events
********************************************

I have latest kernel 6.3.7, compiled from sources. Do I hjave to have some special support for Space Navigator ?
I have compiled in HID and event stuff as per GEntoo how-to page...

"lsusb -v" does list the device as:

***************************************************************************************************************
Bus 003 Device 012: ID 046d:c626 Logitech, Inc. 3Dconnexion Space Navigator 3D Mouse
Device Descriptor:
  bLength                18
  bDescriptorType         1
  bcdUSB               2.00
  bDeviceClass            0 
  bDeviceSubClass         0 
  bDeviceProtocol         0 
  bMaxPacketSize0         8
  idVendor           0x046d Logitech, Inc.
  idProduct          0xc626 3Dconnexion Space Navigator 3D Mouse
  bcdDevice            4.35
  iManufacturer           1 3Dconnexion
  iProduct                2 SpaceNavigator
  iSerial                 0 
  bNumConfigurations      1
  Configuration Descriptor:
    bLength                 9
    bDescriptorType         2
    wTotalLength       0x0029
    bNumInterfaces          1
    bConfigurationValue     1
    iConfiguration          0 
    bmAttributes         0x80
      (Bus Powered)
    MaxPower              100mA
    Interface Descriptor:
      bLength                 9
      bDescriptorType         4
      bInterfaceNumber        0
      bAlternateSetting       0
      bNumEndpoints           2
      bInterfaceClass         3 Human Interface Device
      bInterfaceSubClass      0 
      bInterfaceProtocol      0 
      iInterface              0 
        HID Device Descriptor:
          bLength                 9
          bDescriptorType        33
          bcdHID               1.11
          bCountryCode            0 Not supported
          bNumDescriptors         1
          bDescriptorType        34 Report
          wDescriptorLength     217
         Report Descriptors: 
           ** UNAVAILABLE **
      Endpoint Descriptor:
        bLength                 7
        bDescriptorType         5
        bEndpointAddress     0x81  EP 1 IN
        bmAttributes            3
          Transfer Type            Interrupt
          Synch Type               None
          Usage Type               Data
        wMaxPacketSize     0x0008  1x 8 bytes
        bInterval              10
      Endpoint Descriptor:
        bLength                 7
        bDescriptorType         5
        bEndpointAddress     0x02  EP 2 OUT
        bmAttributes            3
          Transfer Type            Interrupt
          Synch Type               None
          Usage Type               Data
        wMaxPacketSize     0x0008  1x 8 bytes
        bInterval              10
Device Status:     0x0000
  (Bus Powered)
******************************************************************************

Also, "udevadm -u monitor" doesn't show any creation of the device in the /dev, just the map in the /sys:

******************************************
UDEV  [165248.277356] add      /devices/pci0000:00/0000:00:08.1/0000:06:00.4/usb3/3-1/3-1.1 (usb)
UDEV  [165248.278943] bind     /devices/pci0000:00/0000:00:08.1/0000:06:00.4/usb3/3-1/3-1.1 (usb)
******************************************

So I've opened the file "dev" in that map and it lists the char device as 189,268

I tried creating it manually:

"mknod --mode=777 /dev/input/mouse1 b 189 268"

"cat /dev/input/mouse1" doesn't work. I get the response as "non-existing device"

Also, running the daemon manually in verbose mode shows that the daemon doesn't like it either.
It just goes through the same list as as above (+ mouse1 now) and complains in the same way - it didn't recognize any device...

**Comment by Unisa-Merni-392** [comment]
I think I got it.

Good part of the problem is that the daemon fails if usbguard ( USB bus event filter daemon) blocks it after plugin, even if it is allowed later.
But if device is present at usbguard daemon start, by my configuration usbguard daemon allows it and all is fine.

spacenavd daemon finds and recognizes its event device and connects to it. Space Navigator's LED lights up and test_cube example works.

Thanks.

**Reply by jtsiomb** [reply]
Glad to hear you figured it out.

To be clear, the X authentication issue has nothing to do with device detection. If you want to solve that, I'd look into where xwayland puts the authentication cookie, and make XAUTHORITY point to it, or symlink it to your home directory.
